source/solvers/fluid_dynamics_matrix_free_operators.cc

```diff
@@ -1151,44 +1174,90 @@ NavierStokesOperatorBase<dim, number>::do_local_weak_dirichlet_bc(
 
   const auto penalty_parameter = this->effective_beta_face[face_index];
 
-  integrator.evaluate(EvaluationFlags::EvaluationFlags::values |
-                      EvaluationFlags::EvaluationFlags::gradients);
+  if (this->boundary_conditions.type[boundary_index] ==
+      BoundaryConditions::BoundaryType::function_weak)
+    {
+      integrator.evaluate(EvaluationFlags::EvaluationFlags::values |
+                          EvaluationFlags::EvaluationFlags::gradients);
 
-  for (const auto q : integrator.quadrature_point_indices())
+      for (const auto q : integrator.quadrature_point_indices())
+        {
+          typename FEFaceIntegrator::value_type    value_result    = {};
+          typename FEFaceIntegrator::gradient_type gradient_result = {};
+
+          const auto normal_vector = integrator.get_normal_vector(q);
+
+          auto       value    = integrator.get_value(q);
+          const auto gradient = integrator.get_gradient(q);
+
+
+          // If we are assembling the residual, substract the target velocity
+          // from the velocity value.
+          if constexpr (assemble_residual)
+            for (unsigned int d = 0; d < dim; ++d)
+              value[d] -= this->face_target_velocity[face_index][q][d];
```

Comment:
This won't work, because the `face_target_velocity` has only velocity values (dim) and `value` includes also the pressure (dim + 1).  Is there any advantage on doing it like this? If yes, I can simply change the size and add zero for the pressure... however, I find it easier to understand like this...

Reply:
If you think it's better to keep the loop version instead of having this zero padded with a non-existing pressure, it's good for me.